File: Unit2-08/Typescript/Truck.ts
/**
*
* This class creates a truck using methods from vehicles
*
* By:      Evan Beaudoin
* Version: 1.0
* Since:   2024-04-05
*/

import { Vehicle } from './Vehicle'

export class Truck extends Vehicle {
  // Class variables
  private plateNumber: string
  // # of wheels can't be changed once initialized
  private static readonly NUMBER_WHEELS: number = 4

  // Creates the base truck
  public constructor(startColour: string, startMaxSpeed: number, startPlateNumber: string) {
    super(startColour, startMaxSpeed, Truck.NUMBER_WHEELS)
    this.plateNumber = startPlateNumber
  }

  // Sets a new plate number to the truck
  public set plateNumber(newPlateNumber: string) {
    this.plateNumber = newPlateNumber
  }

  // Gets the current plate number
  public get plateNumber() {
    return this.plateNumber
  }

  // Creates a string about the status of the car
  public get status() {
    let statusString = ` -> Speed: ${this.speed}\n`
    statusString += ` -> MaxSpeed: ${this.maxSpeed}\n`
    statusString += ` -> Color: ${this.colour}\n`
    statusString += ` -> License Plate: ${this.plateNumber}\n`
    statusString += ` -> Number of Wheels: ${this.wheels}`

    return statusString
  }

  // Brakes the truck and changes speed
  public break(breakPower: number, breakTime: number, airPressure: number): void {
    this.speed = this.speed - (breakPower * breakTime) - (airPressure * breakTime)
    if (this.speed < 0) {
      this.speed = 0 // Ensure speed doesn't go below zero
    }
  }
}

File: Unit2-08/Java/Truck.java
/*
* This class creates a truck using methods from vehicles
*
* @author  Evan Beaudoin
* @version 1.0
* @since   2024-05-07
*/

/**
 * This is the Truck class.
*/
public class Truck extends Vehicle {
    /**
     * Represents the number of wheels on the vehicle.
     * This constant stores the total number of wheels that the vehicle has.
    */
    private static final int NUMBER_OF_WHEELS = 4;

    /**
     * Variable for the class.
     *
     * @param licensePlateNumber the license plate number on the vehicle
    */
    private String licensePlateNumber = "";

    /**
     * Creates the base truck.
     *
     * @param initialColour the base colour
     * @param initialPlateNumber the base license plate number
     * @param initialMaxSpeed the maximum speed of the vehicle
    */
    public Truck(
            String initialColour,
            double initialMaxSpeed,
            String initialPlateNumber
    ) {
        super(initialColour, initialMaxSpeed, NUMBER_OF_WHEELS);
        this.licensePlateNumber = initialPlateNumber;
    }

    /**
     * Gets the license plate number of the vehicle.
     *
     * @return the current licensePlateNumber
    */
    public String getLicensePlateNumber() {
        return this.licensePlateNumber;
    }

    /**
     * Creates a string about the status of the vehicle.
     *
     * @return the string about the vehicle's status
    */
    public String getStatus() {
        final String breakLine = "\n";
        String status = " -> Speed: " + this.getSpeed() + breakLine;
        status += " -> MaxSpeed: " + this.getMaxSpeed() + breakLine;
        status += " -> Color: " + this.getColour() + breakLine;
        status += " -> License Plate: " + this.licensePlateNumber + breakLine;
        status += " -> Number of Wheels: " + this.getWheels() + breakLine;

        return status;
    }

    /**
     * Sets the license plate of the vehicle.
     *
     * @param newPlateNumber of the vehicle.
    */
    public void setLicensePlateNumber(String newPlateNumber) {
        this.licensePlateNumber = newPlateNumber;
    }

    /**
     * Provides air the the wheels of the truck, which changes speed.
     *
     * @param airPressure applied to the truck
    */
    public void applyAir(double airPressure) {
        this.setSpeed(this.getSpeed() - airPressure / 2);
    }
}
